src/sysupdate/sysupdate-transfer.c

```diff
@@ -105,6 +109,113 @@ Transfer* transfer_new(Context *ctx) {
+                        break;
+                }
+
+                /* First time a UKI is booted it will be recorded with the tries left in efivar, but it will
+                 * have been removed from the filename by bless-boot, so check again. */
+
+                char *e = strrchr(joined, '+');
+                if (!e)
+                        continue;
+                *e = '\0';
+
+                _cleanup_free_ char *truncated = strjoin(joined, ".efi");
+                if (!truncated)
+                        return log_oom_debug();
+
+                r = chase(truncated, *p, /* flags= */ 0, /* ret_path= */ NULL, &fd);
+                if (r >= 0) {
+                        uki = fd_reopen(fd, O_RDONLY|O_CLOEXEC);
+                        if (uki < 0)
+                                return log_debug_errno(errno, "Failed to open UKI file '%s': %m", joined);
+                        break;
+                }
+        }
+        if (uki < 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(ENOENT),
+                                       "Failed to find booted UKI in ESP/XBOOTLDR paths: %m");
+
+        r = inspect_kernel(uki,
+                           /* dir_fd= */ -EBADF,
+                           /* filename= */ NULL,
+                           /* ret_type= */ NULL,
+                           /* ret_cmdline= */ NULL,
+                           /* ret_uname= */ NULL,
+                           /* ret_pretty_name= */ NULL,
+                           &osrel);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to inspect UKI '%s': %m", stub_path);
+        if (!osrel)
+                return log_debug_errno(SYNTHETIC_ERRNO(ENOENT),
+                                       "UKI '%s' does not contain '.osrel' section: %m", stub_path);
+
+        tmproot_fd = mkdtemp_open(/* template= */ NULL, 0, &tmproot);
+        if (tmproot_fd < 0)
+                return log_debug_errno(tmproot_fd, "Failed to create temporary directory for UKI inspection: %m");
+
+        r = mkdirat_parents(tmproot_fd, "usr/lib/os-release", 0755);
+        if (r < 0)
+                return log_debug_errno(errno, "Failed to create directory for os-release file: %m");
+
+        r = write_string_file_at(tmproot_fd, "usr/lib/os-release", osrel, WRITE_STRING_FILE_CREATE);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to write os-release file: %m");
+
+        r = specifier_printf(rvalue, NAME_MAX, specifier_table, tmproot, NULL, &version);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to expand specifiers in ProtectVersion=: %s", rvalue);
+
+        if (!version_is_valid(version))
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "ProtectVersion= string is not valid: %s", version);
+
+        r = strv_extend(ret_protected_versions, version);
```

Comment:
coding style: this is apparently an input + output parameter, not  purely output. hence shouldn't be called ret_xyz